### DISCUSSION TITLE
How to do imports in global scope?

### DISCUSSION THREAD

**Discussion by marius311** [cjdoris/PythonCall.jl#270] [Q&A] [answered]
I'm trying to do something like

```julia
@pyexec """
import numpy as np
"""

@pyexec "np.array([1,2])"
```

Is this possible and if so how?

**Comment by cjdoris** [comment, marked as answer]
`@pyexec` always runs in a local scope. I think putting `global np` at the top of the first call does what you want.

Alternatively `pyexec("import numpy as np", @__module__)` works in global scope unless you provide a local scope in the 3rd arg.

Btw there is no distinction between single- and multi-line code (unlike PyCall's `py"..."` macro). You probably want `@pyeval` for your second call.

**Comment by marius311** [comment]
Got it, thanks. May I suggest making an easier-to-read version of this? Seems like putting a bunch of imports at the top of a script or module is pretty typical. What about this?: 

```julia
@pyexec @__MODULE__ """
import numpy as np
""" 
```

**Reply by cjdoris** [reply]
I used to have essentially this, but removed it because it doesn't really add anything you can't do by other means (using `global np` or `pyexec(...)`) - and I'm trying to make the API smaller not bigger.

**Reply by marius311** [reply]
Its up to you, but fwiw I would just say this is so common with imports at the top of a script / module that its a shame to need the harder-to-read:

```julia
pyexec("""
import numpy as np
""", @__MODULE__)
```

(as well as break the consistency assuming most of your other code is using the macro). 


But if you don't like that, maybe another idea is for the macro to actually enforce Python scope to match Julia scope? A macro [can tell](https://discourse.julialang.org/t/is-there-a-way-to-determine-whether-code-is-toplevel/39939) if its being called in a top-level scope, so its entirely possible to make this work as you'd expect:

```julia

module Foo

@pyexec """
import numpy as np"
global_var = 1
"""

function foo()
    @pyeval "np.array([1,2,3])"
end

end

function bar()
    @pyexec "x = np.array([1,2,3])"
    @pyeval "x"
end

function baz()
    @pyeval "global_var" # OK
    @pyeval "x" # error, even after calling bar()
end

end 
```

schematically, you'd have the macro generate something like:
```julia
if !@isdefined(@__MODULE__.__py_globals__)
    global __py_globals__ = PyDict()
end

if !@istoplevel() && !@isdefined(__py_locals__)
    __py_locals__ = PyDict()
end

pyexec(..., __py_globals__, (@istoplevel() ? __py_globals__ : __py_locals__))
```

Would be more complex if you wanted to handle inner functions, although probably doable.
